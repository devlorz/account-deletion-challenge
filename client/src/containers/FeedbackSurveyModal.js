import _ from 'lodash'
import PropTypes from 'prop-types'
import React from 'react'

import { feedbackSurveyItems } from '../FeedbackSurveyItems'
import Buttons from '../components/Buttons'
import CommentForm from '../components/CommentForm'
import InputForm from '../components/InputForm'

class FeedbackSurveyModal extends React.PureComponent {
  static propTypes = {
    onSubmit: PropTypes.func,
    onBackButton: PropTypes.func,
    title: PropTypes.node,
    showCommentForm: PropTypes.bool,
    comment: PropTypes.string,
    onChangeComment: PropTypes.func,
  }

  constructor(props) {
    super(props)
    this.state = this.setInitialState()
  }

  state = {
    isFocusCommentBox: false,
  }

  setInitialState = () => {
    return _.chain(feedbackSurveyItems)
      .map(item => [item.stack, false])
      .fromPairs()
      .value()
  }

  hasAllUnchecked = () => {
    const FeedbackSurveyItems = this.state
    return (
      _.every(FeedbackSurveyItems, val => val === false) &&
      !this.state.isFocusCommentBox
    )
  }

  onToggleFeedback(stack) {
    this.setState({ [stack]: !this.state[stack] })
  }

  onFocusCommentBox = () => {
    this.setState({ isFocusCommentBox: !this.state.isFocusCommentBox })
  }

  render() {
    return (
      <div>
        <h1>{this.props.title}</h1>
        <div>
          {_.map(feedbackSurveyItems, (item, key) => (
            <div key={key}>
              <label>
                <input
                  type="checkbox"
                  checked={this.state[item.stack]}
                  onClick={() => this.onToggleFeedback(item.stack)}
                />
                {item.title}
              </label>
              <InputForm
                stack={item.stack}
                canComment={item.canComment}
                placeHolder={item.placeHolder}
                isRender={this.state[item.stack]}
              />
            </div>
          ))}
        </div>
        <CommentForm
          isFocusCommentBox={this.state.isFocusCommentBox}
          onChangeComment={this.props.onChangeComment}
          comment={this.props.comment}
          showCommentForm={this.props.showCommentForm}
        />
        <Buttons
          disabled={this.hasAllUnchecked()}
          onBackButton={this.props.onBackButton}
          onSubmit={this.props.onSubmit}
        />
      </div>
    )
  }
}

export default FeedbackSurveyModal
