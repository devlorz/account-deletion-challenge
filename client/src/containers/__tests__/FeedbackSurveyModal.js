import React from 'react'
import { render } from 'react-testing-library'

import FeedbackSurveyModal from '../FeedbackSurveyModal'

test('snapshot', () => {
  const { container } = render(
    <FeedbackSurveyModal
      title="Why would you leave us?"
      onSubmit={jest.fn()}
      onBackButton={jest.fn()}
      showCommentForm
      comment={''}
      onChangeComment={jest.fn()}
    />
  )
  expect(container.firstChild).toMatchSnapshot()
})
