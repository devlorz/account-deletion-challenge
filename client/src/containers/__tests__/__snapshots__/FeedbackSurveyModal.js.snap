// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`snapshot 1`] = `
<div
  data-reactroot=""
>
  <h1>
    Why would you leave us?
  </h1>
  <div>
    <div>
      <label>
        <input
          type="checkbox"
          value="on"
        />
        <!-- react-text: 7 -->
        I do not understand how to use this.
        <!-- /react-text -->
      </label>
      <!-- react-empty: 8 -->
    </div>
    <div>
      <label>
        <input
          type="checkbox"
          value="on"
        />
        <!-- react-text: 12 -->
        I do not need this.
        <!-- /react-text -->
      </label>
      <!-- react-empty: 13 -->
    </div>
    <div>
      <label>
        <input
          type="checkbox"
          value="on"
        />
        <!-- react-text: 17 -->
        I prefer other apps.
        <!-- /react-text -->
      </label>
      <!-- react-empty: 18 -->
    </div>
    <div>
      <label>
        <input
          type="checkbox"
          value="on"
        />
        <!-- react-text: 22 -->
        I cannot find the features that I want.
        <!-- /react-text -->
      </label>
      <!-- react-empty: 23 -->
    </div>
    <div>
      <label>
        <input
          type="checkbox"
          value="on"
        />
        <!-- react-text: 27 -->
        I found this had too many bugs.
        <!-- /react-text -->
      </label>
      <!-- react-empty: 28 -->
    </div>
    <div>
      <label>
        <input
          type="checkbox"
          value="on"
        />
        <!-- react-text: 32 -->
        I found this was too expensive.
        <!-- /react-text -->
      </label>
      <!-- react-empty: 33 -->
    </div>
    <div>
      <label>
        <input
          type="checkbox"
          value="on"
        />
        <!-- react-text: 37 -->
        I found this was too slow.
        <!-- /react-text -->
      </label>
      <!-- react-empty: 38 -->
    </div>
    <div>
      <label>
        <input
          type="checkbox"
          value="on"
        />
        <!-- react-text: 42 -->
        Other reason(s)...
        <!-- /react-text -->
      </label>
      <!-- react-empty: 43 -->
    </div>
  </div>
  <div
    style="margin-top: 2rem;"
  >
    <!-- react-text: 45 -->
    Comments:
    <!-- /react-text -->
    <div>
      <textarea
        name="comment"
        style="border: 1px solid black;"
        type="text"
      />
    </div>
  </div>
  <div>
    <button>
      Back
    </button>
    <button
      disabled=""
    >
      Next
    </button>
  </div>
</div>
`;

exports[`snapshot with no show comment 1`] = `
<div
  data-reactroot=""
>
  <h1>
    Why would you leave us?
  </h1>
  <div>
    <div>
      <label>
        <input
          type="checkbox"
          value="on"
        />
        <!-- react-text: 7 -->
        I do not understand how to use this.
        <!-- /react-text -->
      </label>
      <!-- react-empty: 8 -->
    </div>
    <div>
      <label>
        <input
          type="checkbox"
          value="on"
        />
        <!-- react-text: 12 -->
        I do not need this.
        <!-- /react-text -->
      </label>
      <!-- react-empty: 13 -->
    </div>
    <div>
      <label>
        <input
          type="checkbox"
          value="on"
        />
        <!-- react-text: 17 -->
        I prefer other apps.
        <!-- /react-text -->
      </label>
      <!-- react-empty: 18 -->
    </div>
    <div>
      <label>
        <input
          type="checkbox"
          value="on"
        />
        <!-- react-text: 22 -->
        I cannot find the features that I want.
        <!-- /react-text -->
      </label>
      <!-- react-empty: 23 -->
    </div>
    <div>
      <label>
        <input
          type="checkbox"
          value="on"
        />
        <!-- react-text: 27 -->
        I found this had too many bugs.
        <!-- /react-text -->
      </label>
      <!-- react-empty: 28 -->
    </div>
    <div>
      <label>
        <input
          type="checkbox"
          value="on"
        />
        <!-- react-text: 32 -->
        I found this was too expensive.
        <!-- /react-text -->
      </label>
      <!-- react-empty: 33 -->
    </div>
    <div>
      <label>
        <input
          type="checkbox"
          value="on"
        />
        <!-- react-text: 37 -->
        I found this was too slow.
        <!-- /react-text -->
      </label>
      <!-- react-empty: 38 -->
    </div>
    <div>
      <label>
        <input
          type="checkbox"
          value="on"
        />
        <!-- react-text: 42 -->
        Other reason(s)...
        <!-- /react-text -->
      </label>
      <!-- react-empty: 43 -->
    </div>
  </div>
  <!-- react-empty: 44 -->
  <div>
    <button>
      Back
    </button>
    <button
      disabled=""
    >
      Next
    </button>
  </div>
</div>
`;
